dirtify.py

```diff
@@ -104,37 +87,90 @@
+            tmp2 = cv2.remap(in_img, shrinkMapx, shrinkMapy, cv2.INTER_LINEAR, borderMode=cv2.BORDER_REFLECT)
+            img = cv2.addWeighted(tmp1, 0.5, tmp2, 0.5, 0)
     
     if args.plus:
         # superimpose a white plus at a random position
-        x = round(random.random() * out_w)
-        y = round(random.random() * out_h)
-        cv2.line(dirty, (0,y), (out_w, y), (255,255,255), 1)
-        cv2.line(dirty, (x,0), (x, out_h), (255,255,255), 1)
+        x = round(random.random() * w)
+        y = round(random.random() * h)
+        img = in_img # make a copy so we don't overdraw the clean image
+        cv2.line(img, (0,y), (w, y), (255,255,255), 1)
+        cv2.line(img, (x,0), (x, h), (255,255,255), 1)
     
     if args.xout:
         # Draw a dark red X through the image
-        cv2.line(dirty, (0,0    ), (out_w,out_h), (0,0,64), 2)
-        cv2.line(dirty, (0,out_h), (out_w,0    ), (0,0,64), 2)
+        img = in_img # make a copy so we don't overdraw the clean image
+        cv2.line(img, (0,0), (w,h), (0,0,64), 2)
+        cv2.line(img, (0,h), (w,0), (0,0,64), 2)
     
-    cv2.imwrite(f"{dirty_dir}/{out_name}", dirty)
+    return img
+
+
+# Step through images
+for i, clean in tqdm(enumerate(images), total=len(images)):
+    filename = f"{raw_dir}/{images[i]}"
+    clean = cv2.imread(filename, cv2.IMREAD_COLOR)
+    
+    # resize and crop to make a 64 x 64 image
+    in_h = clean.shape[0]
+    in_w = clean.shape[1]
+    
+    if args.tile:
+        raw_tile_h = 256
+        raw_tile_w = 256
+        for y in range(int(in_h / raw_tile_h)):
+            top = y * raw_tile_h
+            bottom = top + raw_tile_h
+            for x in range(int(in_w / raw_tile_w)):
+                left = x * raw_tile_w
+                right = left + raw_tile_w
+                tile = clean[top:bottom, left:right]
+                
+                # Scale from raw tile size to output size
+                scale = out_h / raw_tile_h
+                tile = cv2.resize(tile, None, fx = scale, fy = scale, interpolation = cv2.INTER_AREA)
+
+                # Add that dirt!
+                dirty = add_dirt(tile, out_w, out_h)
+                
+                # And save files
+                out_name = os.path.splitext(images[i])[0] + f" ({x},{y}).png" # Change the extension to png (for lossless image)
+                cv2.imwrite(f"{clean_dir}/{out_name}", tile)
+                cv2.imwrite(f"{dirty_dir}/{out_name}", dirty)
+
+    else:
+        scale = max(out_h / in_h, out_w / in_w)
+
+        # resize and crop
+        mid_y = scale * in_h / 2
+        mid_x = scale * in_w / 2
+        clean = cv2.resize(clean, None, fx = scale, fy = scale, interpolation = cv2.INTER_AREA)
+        clean = clean[round(mid_y - out_h/2):round(mid_y + out_h/2), round(mid_x - out_w/2):round(mid_x + out_w/2)]
+
+        # Add that dirt!
+        dirty = add_dirt(clean, out_w, out_h)
```

Comment:
If we need to make a copy of `clean`, let's do it before we call `add_dirt()` and pass that copy to `add_dirt()`, instead of making potentially several copies within `add_dirt()`.

However, I'm not sure we need to make a copy at all. We only use the in-memory clean image once, correct? If so, we can write it out (`cv2.imwrite(f"{clean_dir}/{out_name}", clean)`) **before** calling `add_dirt()`, and then pass it into `add_dirt()` and write out the dirty version when we're finished. Or something similar.

Reply:
Resolved.